cpp/src_prims/stats/histogram.h

```diff
@@ -0,0 +1,488 @@
+void smemWarpHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
+                  int tpb, cudaStream_t stream) {
+  int nblks = ceildiv<int>(VecLen ? n / VecLen : n, tpb);
+  size_t smemSize = nbins * sizeof(int);
+  smemWarpHistKernel<DataT, BinnerOp, IdxT, VecLen>
+    <<<nblks, tpb, smemSize, stream>>>(bins, data, n, nbins, op);
+  CUDA_CHECK(cudaGetLastError());
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+void histogramVecLen(HistType type, int* bins, IdxT nbins, const DataT* data,
+                     IdxT n, int tpb, cudaStream_t stream, BinnerOp op) {
+  CUDA_CHECK(cudaMemsetAsync(bins, 0, nbins * sizeof(int), stream));
+  switch (type) {
+    case HistTypeGmemWarp:
+      gmemWarpHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                                  stream);
+      break;
+    case HistTypeGmem:
+      gmemHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                              stream);
+      break;
+    case HistTypeSmemWarp:
+      smemWarpHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                                  stream);
+      break;
+    case HistTypeSmem:
+      smemHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                              stream);
+      break;
+    case HistTypeSmemBits16:
+      smemBitsHist<DataT, BinnerOp, IdxT, 16, VecLen>(bins, nbins, data, n, op,
+                                                      tpb, stream);
+      break;
+    case HistTypeSmemBits8:
+      smemBitsHist<DataT, BinnerOp, IdxT, 8, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemBits4:
+      smemBitsHist<DataT, BinnerOp, IdxT, 4, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemBits2:
+      smemBitsHist<DataT, BinnerOp, IdxT, 2, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemBits1:
+      smemBitsHist<DataT, BinnerOp, IdxT, 1, VecLen>(bins, nbins, data, n, op,
+                                                     tpb, stream);
+      break;
+    case HistTypeSmemHash:
+      smemHashHist<DataT, BinnerOp, IdxT, VecLen>(bins, nbins, data, n, op, tpb,
+                                                  stream);
+      break;
+    default:
+      ASSERT(false, "histogram: Invalid type passed '%d'!", type);
+  };
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT>
+void histogramImpl(HistType type, int* bins, IdxT nbins, const DataT* data,
+                   IdxT n, int tpb, cudaStream_t stream, BinnerOp op) {
+  size_t bytes = n * sizeof(DataT);
+  if (n <= 0) return;
+  if (16 / sizeof(DataT) && bytes % 16 == 0) {
```

Comment:
`if (sizeof(DataT) <= 16 && bytes % 16 == 0)`

The same applies to other comparisons here.

Reply:
good catch. The correct check here should have been:
`if (16 % sizeof(DataT) == 0 && bytes % 16 == 0) {`
Have fixed this one. Thanks